crates/sui-indexer/src/handlers/pruner.rs

```diff
@@ -38,61 +41,52 @@ impl Pruner {
     }
 
     pub async fn start(&self, cancel: CancellationToken) -> IndexerResult<()> {
-        let mut last_seen_max_epoch = 0;
+        // Spawn a separate task to continuously update the watermarks for the reader. We can't
+        // reliably update watermarks while we're in the middle of a pruning operation. This is
+        // because the pruning operation itself may take a considerable amount of time, during which
+        // the system state could change significantly. Pruner prunes on the main thread to limit
+        // concurrency.
```

Comment:
Not sure I follow this explanation. My naïve assumption was that the pruner updates `pruner_lo` in the same DB transaction that actually does the pruning, because otherwise there could be a discrepancy between the actually pruned region, and the region recorded by the watermark table, in the crash recovery scenario.

What's the scenario in which the system state changes and it causes problems?

Reply:
> the pruner updates pruner_lo in the same DB transaction that actually does the pruning

Yes that makes sense. 

> What's the scenario in which the system state changes and it causes problems?

While we prune `transactions` lookup tables, a new epoch comes in, committer updates `hi`, and `pruner` sees that we've exceeded retention. So pruner would update `epoch_lo`, `epoch_hi`, and `timestamp_ms`. During this time, the pruner should finish its current work in pruning lookup tables, and can pick up the new work when it completes.